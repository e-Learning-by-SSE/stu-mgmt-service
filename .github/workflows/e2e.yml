name: E2E Tests
on: push

jobs:
  # Label of the runner job
  runner-job:
    # You must use a Linux environment when using service containers or container jobs
    runs-on: ubuntu-latest

    # Service containers to run with `runner-job`
    services:
      # Label used to access the service container
      postgres:
        # Docker Hub image
        image: postgres:latest
        env:
          POSTGRES_DB: StudentMgmtDb
          POSTGRES_USER: postgres
          POSTGRES_PASSWORD: admin
        # Set health checks to wait until postgres has started
        options: >-
          --health-cmd pg_isready
          --health-interval 10s
          --health-timeout 5s
          --health-retries 5
        ports:
          - 5432:5432

    steps:
      # Downloads a copy of the code in your repository before running CI tests
      - name: Check out repository code
        uses: actions/checkout@v2

      # Performs a clean installation of all dependencies in the `package.json` file
      # For more information, see https://docs.npmjs.com/cli/ci.html
      - name: Install dependencies
        run: npm ci

      - name: Run e2e tests
        run: npm run test:e2e
