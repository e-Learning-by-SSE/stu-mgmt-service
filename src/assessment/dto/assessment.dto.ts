import { ApiProperty, ApiPropertyOptional, OmitType } from "@nestjs/swagger";
import { AssignmentDto } from "../../course/dto/assignment/assignment.dto";
import { ParticipantDto } from "../../course/dto/course-participant/participant.dto";
import { GroupDto } from "../../course/dto/group/group.dto";
import { UserDto } from "../../shared/dto/user.dto";
import { UserId } from "../../shared/entities/user.entity";
import { PartialAssessmentDto } from "./partial-assessment.dto";

export class AssessmentDto {
	/** Unique identifier of this assessment. */
	@ApiProperty({ description: "Unique identifier of this assessment." })
	id: string;

	/** Identifier of the assignment that is being evaluated by this assessment. */
	@ApiProperty({
		description: "Identifier of the assignment that is being evaluated by this assessment."
	})
	assignmentId: string;

	assignment?: AssignmentDto;

	@ApiProperty({ description: "Marks the assessment as a draft." })
	isDraft: boolean;

	/** The amount of points that the student or group achieved with their submission. */
	@ApiProperty({
		description:
			"The amount of points that the student or group achieved with their submission."
	})
	achievedPoints?: number;

	/** A comment providing additional feedback. */
	@ApiPropertyOptional({ description: "A comment providing additional feedback." })
	comment?: string;

	/** If a group submission is being evaluated, contains the identifier of the group. */
	@ApiPropertyOptional({
		description:
			"If a group submission is being evaluated, contains the identifier of the group."
	})
	groupId?: string;

	/** The group, whose submission was evaluated by this assessment. */
	group?: GroupDto;

	/** If a single user is being evaluated, contains the identifier of the user. */
	@ApiPropertyOptional({
		description: "If a single user is being evaluated, contains the identifier of the user."
	})
	userId?: string;

	/** If assessment targets a single user, contains the user. */
	participant?: ParticipantDto;

	/** Identifier of the creator of this assessment. */
	@ApiPropertyOptional({ description: "Identifier of the creator of this assessment." })
	creatorId: string;

	/** The creator of this assessment. */
	creator?: UserDto;

	@ApiPropertyOptional({
		description: "Identifier of the last user that updated this assessment."
	})
	lastUpdatedById?: UserId;
	lastUpdatedBy?: UserDto;

	creationDate?: Date;
	updateDate?: Date;

	partialAssessments?: PartialAssessmentDto[];
}

/** Version of AssessmentDto containing only properties that can used for creation. */
export class AssessmentCreateDto extends OmitType(AssessmentDto, [
	"group",
	"participant",
	"assignment",
	"creator",
	"creatorId",
	"id",
	"lastUpdatedBy",
	"lastUpdatedById",
	"creationDate",
	"updateDate"
]) {}
/** Version of AssessmentDto containing only properties that can be updated. */
export class AssessmentUpdateDto {
	achievedPoints?: number;
	isDraft?: boolean;
	comment?: string;
	partialAssessments?: PartialAssessmentDto[];
}
