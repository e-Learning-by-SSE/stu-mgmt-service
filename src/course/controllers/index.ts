import { ActivityController } from "../../activity/activity.controller";
import { SubmissionController } from "../../submission/submission.controller";
import { AssignmentRegistrationController } from "./assignment-registration.controller";
import { AssignmentController } from "./assignment.controller";
import { CourseConfigController } from "./config.controller";
import { CourseParticipantsController } from "./course-participants.controller";
import { CourseController } from "./course.controller";
import { GroupController } from "./group.controller";

export const Controllers = [
	AssignmentController,
	AssignmentRegistrationController,
	CourseConfigController,
	CourseParticipantsController,
	CourseController,
	GroupController,
	ActivityController,
	SubmissionController
];
