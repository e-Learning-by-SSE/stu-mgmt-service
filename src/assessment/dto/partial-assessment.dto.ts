import { IsNotEmpty } from "class-validator";
import { MarkerDto } from "./marker.dto";

export class PartialAssessmentDto {
	/**
	 * Unique identifier for this partial assessment.
	 * Must be unique within a single assessment.
	 * If not specified, it will be automatically generated.
	 * @example "task_1"
	 */
	key?: string;
	@IsNotEmpty()
	title: string;
	/** Determines whether the partial assessment should be included, when the assessment is not a draft anymore. */
	draftOnly: boolean;
	comment?: string;
	points?: number;
	markers?: MarkerDto[];
}
