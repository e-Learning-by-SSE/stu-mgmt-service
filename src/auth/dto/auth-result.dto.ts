import { UserDto } from "../../shared/dto/user.dto";

export class AuthResultDto {
	user: UserDto;
	accessToken: string;
	expiration?: Date;
}
