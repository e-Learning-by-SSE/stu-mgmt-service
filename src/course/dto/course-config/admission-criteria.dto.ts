import { IsArray } from "class-validator";
import { AdmissionRuleDto } from "../../../admission-status/dto/admission-rule.dto";

export class AdmissionCriteriaDto {
	@IsArray()
	rules: AdmissionRuleDto[];
}
