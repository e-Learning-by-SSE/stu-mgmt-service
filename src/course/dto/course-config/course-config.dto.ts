import { ApiPropertyOptional, PartialType, PickType } from "@nestjs/swagger";
import { AdmissionCriteriaDto } from "./admission-criteria.dto";
import { GroupSettingsDto } from "./group-settings.dto";

/**
 * A dto that contains the configuration of a course.
 */
export class CourseConfigDto {
	@ApiPropertyOptional({ description: "Unique identifier of the configuration." })
	id?: number;
	groupSettings?: GroupSettingsDto;
	admissionCriteria?: AdmissionCriteriaDto;

	/** Password required to sign up for the course. */
	@ApiPropertyOptional({ description: "Password required to sign up for the course." })
	password?: string;
}

/** Version of CourseConfigDto that only contains editable properties. */
export class CourseConfigUpdateDto extends PartialType(PickType(CourseConfigDto, ["password"])) {}
