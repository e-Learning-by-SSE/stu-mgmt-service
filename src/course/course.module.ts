import { HttpModule } from "@nestjs/axios";
import { Module } from "@nestjs/common";
import { CqrsModule } from "@nestjs/cqrs";
import { TypeOrmModule } from "@nestjs/typeorm";
import { AssessmentRepository } from "../assessment/repositories/assessment.repository";
import { AuthModule } from "../auth/auth.module";
import { UserRepository } from "../user/repositories/user.repository";
import { JoinRandomGroupHandler } from "./commands/join-random-group.handler";
import { Controllers } from "./controllers";
import { GroupRegistrationRelation } from "./entities/group-registration-relation.entity";
import { UserJoinedGroupHandler } from "./events/group/user-joined-group.event";
import {
	CloseEmptyGroupsHandler,
	UserLeftGroupHandler
} from "./events/group/user-left-group.event";
import { Guards } from "./guards";
import { CourseMemberGuard } from "./guards/course-member/course-member.guard";
import { ParticipantIdentityGuard } from "./guards/identity.guard";
import { TeachingStaffGuard } from "./guards/teaching-staff.guard";
import { QueryHandlers } from "./queries";
import { Repositories } from "./repositories";
import { Services } from "./services";

@Module({
	imports: [
		TypeOrmModule.forFeature([
			...Repositories,
			UserRepository,
			GroupRegistrationRelation,
			AssessmentRepository
		]),
		CqrsModule,
		HttpModule,
		AuthModule
	],
	controllers: [...Controllers],
	providers: [
		...Services,
		...Guards,
		...QueryHandlers,
		JoinRandomGroupHandler,
		UserJoinedGroupHandler,
		UserLeftGroupHandler,
		CloseEmptyGroupsHandler
	],
	exports: [
		TypeOrmModule,
		CourseMemberGuard,
		TeachingStaffGuard,
		ParticipantIdentityGuard,
		...Services
	]
})
export class CourseModule {}
