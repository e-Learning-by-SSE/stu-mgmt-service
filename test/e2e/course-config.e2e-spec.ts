import { AdmissionCriteriaDto } from "../../src/course/dto/course-config/admission-criteria.dto";
import { AssignmentTemplateDto } from "../../src/course/dto/course-config/assignment-template.dto";
import {
	CourseConfigDto,
	CourseConfigUpdateDto
} from "../../src/course/dto/course-config/course-config.dto";
import {
	GroupSettingsDto,
	GroupSettingsUpdateDto
} from "../../src/course/dto/course-config/group-settings.dto";
import { CourseDto } from "../../src/course/dto/course/course.dto";
import { ADMISSION_CRITERIA_MOCK } from "../mocks/course-config/admission-criteria.mock";
import { COURSE_CONFIG_JAVA_1920 } from "../mocks/course-config/course-config.mock";
import { GROUP_SETTINGS_GROUPS_ALLOWED_MIN2_MAX3_SELF } from "../mocks/course-config/group-settings.mock";
import { COURSE_JAVA_1920 } from "../mocks/courses.mock";
import { USER_STUDENT_2_JAVA, USER_STUDENT_JAVA } from "../mocks/users.mock";
import { TestSetup } from "../utils/e2e";
import { copy } from "../utils/object-helper";

const course: CourseDto = COURSE_JAVA_1920;

describe("CourseConfig E2E", () => {
	let setup: TestSetup;

	beforeAll(async () => {
		setup = await TestSetup.create();
	});

	afterAll(async () => {
		await setup.teardown();
	});

	describe("POST-REQUEST of CourseConfigController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createCourses();
		});

		it("(POST) .../config Creates configuration and returns it", () => {
			const config = copy(COURSE_CONFIG_JAVA_1920);
			const expected = COURSE_CONFIG_JAVA_1920;

			return setup
				.request()
				.post(`/courses/${course.id}/config`)
				.send(config)
				.expect(201)
				.expect(({ body }) => {
					const result = body as CourseConfigDto;
					expect(result).toEqual(expected);
				});
		});

		it("(POST) .../config/admission-criteria Creates admission criteria and returns it", async () => {
			await setup.dbMockService.createCourseConfig();
			const config = copy(COURSE_CONFIG_JAVA_1920);
			const criteria = copy(ADMISSION_CRITERIA_MOCK);
			const expected = ADMISSION_CRITERIA_MOCK;

			return setup
				.request()
				.post(`/courses/${course.id}/config/${config.id}/admission-criteria`)
				.send(criteria)
				.expect(201)
				.expect(({ body }) => {
					const result = body as AdmissionCriteriaDto;
					expect(result).toEqual(expected);
				});
		});
	});

	describe("PUT-REQUEST of CourseConfigController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		describe("(PUT) courses/{courseId}/config/admission-from-previous-semester", () => {
			it("With matrNrs -> Returns DTO", () => {
				const matrNrs = [USER_STUDENT_JAVA.matrNr, USER_STUDENT_2_JAVA.matrNr];

				return setup
					.request()
					.put(`/courses/${course.id}/config/admission-from-previous-semester`)
					.send(matrNrs)
					.expect(200)
					.expect(({ body }) => {
						expect(body).toMatchSnapshot();
					});
			});

			it("Empty array -> Clears existing data", () => {
				const matrNrs = [];

				return setup
					.request()
					.put(`/courses/${course.id}/config/admission-from-previous-semester`)
					.send(matrNrs)
					.expect(200)
					.expect(({ body }) => {
						expect(body).toMatchSnapshot();
					});
			});

			it("No payload -> Returns status code 400", () => {
				return setup
					.request()
					.put(`/courses/${course.id}/config/admission-from-previous-semester`)
					.expect(400);
			});
		});
	});

	describe("GET-REQUESTS of CourseConfigConroller (e2e)", () => {
		beforeAll(async () => {
			await setup.clearDb();
			await setup.dbMockService.createAll();
		});

		it("(GET) .../config Returns complete course configuration", () => {
			const expected = COURSE_CONFIG_JAVA_1920;

			return setup
				.request()
				.get(`/courses/${course.id}/config`)
				.expect(200)
				.expect(({ body }) => {
					const result = body as CourseConfigDto;
					expect(result).toEqual(expected);
				});
		});

		it("(GET) .../config/group-settings Returns group settings", () => {
			const expected = COURSE_CONFIG_JAVA_1920.groupSettings;

			return setup
				.request()
				.get(`/courses/${course.id}/config/group-settings`)
				.expect(200)
				.expect(({ body }) => {
					const result = body as GroupSettingsDto;
					expect(result).toEqual(expected);
				});
		});

		it("(GET) .../config/admission-criteria Returns admission criteria", () => {
			const expected = COURSE_CONFIG_JAVA_1920.admissionCriteria;

			return setup
				.request()
				.get(`/courses/${course.id}/config/admission-criteria`)
				.expect(200)
				.expect(({ body }) => {
					const result = body as AdmissionCriteriaDto;
					expect(result).toEqual(expected);
				});
		});

		it("(GET) .../config/assignment-templates Returns assignment templates", () => {
			const expected = COURSE_CONFIG_JAVA_1920.assignmentTemplates;

			return setup
				.request()
				.get(`/courses/${course.id}/config/assignment-templates`)
				.expect(200)
				.expect(({ body }) => {
					const result = body as AssignmentTemplateDto[];
					expect(result).toEqual(expected);
				});
		});
	});

	describe("PATCH-REQUEST of CourseConfigController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createCourses();
			await setup.dbMockService.createCourseConfig();
			await setup.dbMockService.createGroupSettings();
			await setup.dbMockService.createAdmissionCriteria();
			await setup.dbMockService.createAssignmentTemplates();
		});

		it("(PATCH) .../config Updates course config (no relations)", () => {
			const expected = copy(COURSE_CONFIG_JAVA_1920);
			expected.password = "newPassword";

			const update: CourseConfigUpdateDto = {
				password: expected.password
			};

			return setup
				.request()
				.patch(`/courses/${course.id}/config`)
				.send(update)
				.expect(200)
				.expect(({ body }) => {
					const result = body as CourseConfigDto;
					expect(result).toEqual(expected);
				});
		});

		it("(PATCH) .../config/group-settings Updates group settings", () => {
			const groupSettings = copy(GROUP_SETTINGS_GROUPS_ALLOWED_MIN2_MAX3_SELF);
			groupSettings.nameSchema += "Update";
			groupSettings.allowGroups = !groupSettings.allowGroups;
			groupSettings.selfmanaged = !groupSettings.selfmanaged;
			groupSettings.sizeMin += 1;
			groupSettings.sizeMax += 1;

			const update: GroupSettingsUpdateDto = {
				nameSchema: groupSettings.nameSchema,
				allowGroups: groupSettings.allowGroups,
				selfmanaged: groupSettings.selfmanaged,
				sizeMin: groupSettings.sizeMin,
				sizeMax: groupSettings.sizeMax
			};

			return setup
				.request()
				.patch(`/courses/${course.id}/config/group-settings`)
				.send(update)
				.expect(200)
				.expect(({ body }) => {
					const result = body as GroupSettingsDto;
					expect(result).toEqual(groupSettings);
				});
		});

		it("(PATCH) .../config/admission-criteria Updates admission criteria", () => {
			const criteria = copy(ADMISSION_CRITERIA_MOCK);
			criteria.rules[0].requiredPercent = 55; // Change required percent
			criteria.rules = [criteria.rules[0], criteria.rules[1]]; // Remove one rule

			return setup
				.request()
				.patch(`/courses/${course.id}/config/admission-criteria`)
				.send(criteria)
				.expect(200)
				.expect(({ body }) => {
					const result = body as AdmissionCriteriaDto;
					expect(result).toEqual(criteria);
				});
		});
	});

	describe("DELETE-REQUEST of CourseConfigController (e2e)", () => {
		beforeEach(async () => {
			await setup.clearDb();
			await setup.dbMockService.createCourses();
			await setup.dbMockService.createCourseConfig();
			await setup.dbMockService.createGroupSettings();
			await setup.dbMockService.createAdmissionCriteria();
			await setup.dbMockService.createAssignmentTemplates();
		});

		it("(DELETE) .../config Deletes complete config", () => {
			return setup.request().delete(`/courses/${course.id}/config/`).expect(200);
		});

		it("(DELETE) .../config/admission-criteria Deletes admission criteria ", () => {
			return setup
				.request()
				.delete(`/courses/${course.id}/config/admission-criteria`)
				.expect(200);
		});
	});
});
