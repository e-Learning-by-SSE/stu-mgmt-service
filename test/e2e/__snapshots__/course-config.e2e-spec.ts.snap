// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseConfig E2E PUT-REQUEST of CourseConfigController (e2e) (PUT) courses/{courseId}/config/admission-from-previous-semester Empty array -> Clears existing data 1`] = `
Object {
  "matrNrs": Array [],
  "participants": Array [],
}
`;

exports[`CourseConfig E2E PUT-REQUEST of CourseConfigController (e2e) (PUT) courses/{courseId}/config/admission-from-previous-semester With matrNrs -> Returns DTO 1`] = `
Object {
  "matrNrs": Array [
    123456,
    111111,
  ],
  "participants": Array [
    Object {
      "displayName": "Max Mustermann",
      "email": "max.mustermann@test.com",
      "group": Object {
        "hasPassword": true,
        "id": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
        "isClosed": false,
        "name": "Testgroup 1",
      },
      "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
      "matrNr": 123456,
      "role": "STUDENT",
      "userId": "a019ea22-5194-4b83-8d31-0de0dc9bca53",
      "username": "mmustermann",
    },
    Object {
      "displayName": "Hans Peter",
      "email": "hans.peter@test.com",
      "group": Object {
        "hasPassword": true,
        "id": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
        "isClosed": false,
        "name": "Testgroup 1",
      },
      "groupId": "b4f24e81-dfa4-4641-af80-8e34e02d9c4a",
      "matrNr": 111111,
      "role": "STUDENT",
      "userId": "40f59aad-7473-4455-a3ea-1214f19b2565",
      "username": "hpeter",
    },
  ],
}
`;
